arch/arm/core/swap.S

```diff
@@ -324,6 +352,47 @@ _oops:
     blx _do_kernel_oops
     pop {pc}
 
+#if CONFIG_USERSPACE
+    /*
+     * System call will setup a jump to the _do_arm_syscall function
+     * when the SVC returns via the bx lr.
+     *
+     * There is some trickery involved here because we have to preserve
+     * the original LR value so that we can return back to the caller of
+     * the SVC.
+     *
+     * On SVC exeption, the stack looks like the following:
+     * r0 - r1 - r2 - r3 - r12 - LR - PC - PSR
+     * r5 - r6 - call id - saved LR
+     *
+     */
+_do_syscall:
+    ldr r1, [r0, #24]   /* grab address of PC from stack frame */
+    str r1, [r0, #44]   /* store address to use for LR after syscall */
+    ldr r1, =_arm_do_syscall
+    str r1, [r0, #24]   /* overwrite the LR to point to _arm_do_syscall */
```

Comment:
So, nothing else can overwrite this in order to redirect control flow after privilege elevation?  We're operating on the unprivileged stack but relying on the fact that no other thread can access it?  Why do we refer to [r0, #24] as PC above but LR here?

Reply:
I should fix this comment.  What we are doing is storing off the PC that would have been used in the return into a LR storage spot we set aside.  Then we are overwriting the PC to return from interrupt into the _arm_do_syscall.

Reply:
So, to be clear, the security of this depends on the fact that no other thread can write to this thread's (user) stack and there is no way that the overwritten PC on the stack frame could be clobbered with a different value prior to the bx lr?

Reply:
At the point that this is running, we are in exception mode doing a SVC.  Only a higher priority interrupt could get inbetween us making this modification and the bx lr which pops all this information off the stack and returns to the new PC.

Reply:
SMP Zephyr is going to be a reality soon. Is there any multicore ARM SoC where this would be an issue?